### DISCUSSION TITLE
Types are here!

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#112] [Announcements]
Update (1.1.0-beta.86) adds a [`~/.kit/kit.d.ts`](https://github.com/johnlindquist/kit/blob/main/kit.d.ts) to allow better code hinting and completion.

❗️After updating, you will need to manually "link" your `~/.kenv` to your `~/.kit` for the benefits (This will happen automatically for new users during install)

Method 1 - Install and run this script

[Click to install link-kit](https://scriptkit.app/api/new?name=link-kit&url=https://gist.githubusercontent.com/johnlindquist/f238cb1b3a3ed97890657ccf154d12b1/raw/a488a8b6c331d527bb0433a6b8df9428263b85a0/link-kit.js)

```js
await cli("install", "~/.kit")
```

Method  2 - In your terminal
```bash
PATH=~/.kit/node/bin ~/.kit/node/bin/npm --prefix ~/.kenv i ~/.kit
```

Now your scripts in your `~/.kenv/scripts` should have completion/hinting for globals included in the "preloaded" scripts.

> I still need to add types for the helpers that load scripts from dirs `kit()`, `cli()`, etc.

Please let me know how it goes and if you have any questions. Thanks!

**Comment by johnlindquist** [comment]
Oh yeah, pro tip for the `npm` helper. Adding this jsdoc comment will give you code-completion when using `npm`:

```js
/** @type {typeof import("query-string")} */
let qs = await npm("query-string")
```
I'm thinking through the best way to help with helpers/snippets/etc to make this smoother

> Technical note: `npm` is impossible to "type" because it wraps `import` which requires a "string literal" (not a variable)

**Reply by horacioh** [reply]
FYI if the package is a default export you can add `.default` at the end of the comment to get the types!

```js
/** @type {typeof import("remark-parse").default} */
let remarkParse = await npm("remark-parse")
```

🎉

**Comment by zzeleznick** [comment]
Method 2 worked for me and after installing VS Code (sublime is my default editor) I was able to see the [code hinting](https://code.visualstudio.com/docs/editor/intellisense) in action!

<img width="628" alt="type-hinting" src="https://user-images.githubusercontent.com/5779832/113490518-599ab280-947f-11eb-93bf-c8e3d81a7d9f.png">

**Reply by johnlindquist** [reply]
Does sublime not pick up on type defs from installed packages? I'll need to get around to testing it out.

**Reply by zzeleznick** [reply]
I might be missing a package or two, but if you find out how to set it up in Sublime that would be very helpful!

**Comment by jsjoeio** [comment]
Method 2 worked for me as well 

<img width="434" alt="image" src="https://user-images.githubusercontent.com/3806031/113490610-1987ff80-9480-11eb-9436-8bdf4385ca40.png">

Would be nice if Method 1 was a "Install Script" kit link like the ones on [here](https://scriptkit.app/scripts/johnlindquist) .

**Reply by johnlindquist** [reply]
Updated Method 1 with an install link 👍

**Reply by jsjoeio** [reply]
New install link worked like a charm!


https://user-images.githubusercontent.com/3806031/113513176-21e24800-951d-11eb-8d67-debe540a2ed4.mov

**Reply by johnlindquist** [reply]
Thanks. I was really tempted to do it automagically, but I promised myself never to read/alter/etc someone's ~/.kenv after the initial setup.

**Comment by MinatoHikari** [comment]
how to use types package like @types/node?

**Reply by johnlindquist** [reply]
If you're using VS Code, then you can treat ~/.kenv like any other node project. npm install and use whatever types you like.
